[JSC] Lower ValueBitXXX into ArithBitXXX when AI proved input are Int32

Constellation opened the pull request:
#### 5afa83f749d015298572502c49d8a2fe65d141c1
<pre>
[JSC] Lower ValueBitXXX into ArithBitXXX when AI proved input are Int32
<a href="https://bugs.webkit.org/show_bug.cgi?id=280916">https://bugs.webkit.org/show_bug.cgi?id=280916</a>
<a href="https://rdar.apple.com/137314028">rdar://137314028</a>

Reviewed by Keith Miller.

We found that in some cases, we failed to track accurate types in
prediction, and end up using conservative node.

    if (typeof x === &apos;number&apos;) {
        x | 0 // Depending on the use of x, but the prediction can be String|Int32 still.
    } else {
        x + &quot;42&quot;
    }

And passing String|Int32 to x, end up generating ValueBitOr. But when
inlining them and running AI, we may prove that x is always Int32 later.

In this patch, we attempt to lower these cases into more optimized
nodes, like ValueBitOr =&gt; ArithBitOr when we prove that both input are
definitely Int32. We also expand CompareEq etc.&apos;s coverage onto the
proven types and some more additional handling in ArithBitOr in AI.

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):

Canonical link: <a href="https://commits.webkit.org/284721@main">https://commits.webkit.org/284721@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea875440ee2ad6fe82a7f172687506afe56deb1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74403 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57520 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21332 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45242 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60614 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36182 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41901 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18049 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19853 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/63435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18393 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76122 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69561 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17632 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14582 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63377 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11414 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5041 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91343 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10760 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45522 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/289 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/19910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46596 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47873 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -631,11 +637,18 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+            ASSERT(node->child1().useKind() != UntypedUse);
+            if (node->child2()->isInt32Constant() && !node->child2()->asInt32()) {
+                m_state.setShouldTryConstantFolding(true);
+                forNode(node) = forNode(node->child1());
```
`x & 0` is always 0, right? So, should it be this?
```
forNode(node) = forNode(node->child2());
```

Constellation replied:
Changed.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1353,6 +1353,56 @@ class ConstantFoldingPhase : public Phase {
+                    m_insertionSet.insertCheck(m_graph, indexInBlock, node);
+                    node->convertToIdentityOn(node->child1().node());
+                    changed = true;
+                    break;
```
This `break` is redundant.

Constellation replied:
This is kind of early return and we would like to have that so that we can easily place more optimization after this if-clause.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -631,11 +637,18 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
-                isBoolInt32Speculation(forNode(node->child2()).m_type))) {
-            setNonCellTypeForNode(node, SpecBoolInt32);
-            break;
+        if (node->op() == ArithBitAnd) {
```
Should introduce similar things for `ArithBitOr` and `ArithBitXor`?

```
if (node->op() == ArithBitOr) {
    ASSERT(node->child1().useKind() != UntypedUse);
    if (node->child2()->isInt32Constant() && !node->child2()->asInt32()) {
        // x | 0 -> x, so fold to child1
        m_state.setShouldTryConstantFolding(true);
        forNode(node) = forNode(node->child1());
        break;
    }
    
    if (isBoolInt32Speculation(forNode(node->child1()).m_type) || isBoolInt32Speculation(forNode(node->child2()).m_type)) {
        setNonCellTypeForNode(node, SpecBoolInt32);
        break;
    }
    
    setNonCellTypeForNode(node, SpecInt32Only);
}
```

and

```
if (node->op() == ArithBitXor) {
    ASSERT(node->child1().useKind() != UntypedUse);
    
    // Case 1: x ^ 0 -> x (Identity for XOR)
    if (node->child2()->isInt32Constant() && !node->child2()->asInt32()) {
        // If child2 is the constant 0, simplify to child1
        m_state.setShouldTryConstantFolding(true);
        forNode(node) = forNode(node->child1());
        break;
    }
    
    // Case 2: x ^ x -> 0 (XOR with itself results in 0)
    if (node->child1() == node->child2()) {
        // If child1 and child2 are the same, the result is always 0
        m_state.setShouldTryConstantFolding(true);
        setConstant(node, jsNumber(0));
        break;
    }

    if (isBoolInt32Speculation(forNode(node->child1()).m_type) || isBoolInt32Speculation(forNode(node->child2()).m_type)) {
        setNonCellTypeForNode(node, SpecBoolInt32);
        break;
    }

    setNonCellTypeForNode(node, SpecInt32Only);
}
```

Constellation replied:
```
    if (isBoolInt32Speculation(forNode(node->child1()).m_type) || isBoolInt32Speculation(forNode(node->child2()).m_type)) {
        setNonCellTypeForNode(node, SpecBoolInt32);
        break;
    }
    ```
    This is only applicable to ArithBitAnd.
Added the other part.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1353,6 +1353,56 @@ class ConstantFoldingPhase : public Phase {
                 break;
             }
 
+            case ArithBitOr: {
```
I am thinking we should do similar things for all of them in this patch or other patches:
```
ArithBitXor:
x ^ 0 -> x (identity operation).
x ^ x -> 0 (self-XOR results in 0).

ArithBitAnd:
x & 0 -> 0 (AND with 0 results in 0).
x & -1 -> x (AND with -1 is identity).

ArithBitOr:
x | 0 -> x (OR with 0 is identity).
```

Constellation committed:
[JSC] Lower ValueBitXXX into ArithBitXXX when AI proved input are Int32
https://bugs.webkit.org/show_bug.cgi?id=280916
rdar://137314028

Reviewed by Keith Miller.

We found that in some cases, we failed to track accurate types in
prediction, and end up using conservative node.

    if (typeof x === 'number') {
        x | 0 // Depending on the use of x, but the prediction can be String|Int32 still.
    } else {
        x + "42"
    }

And passing String|Int32 to x, end up generating ValueBitOr. But when
inlining them and running AI, we may prove that x is always Int32 later.

In this patch, we attempt to lower these cases into more optimized
nodes, like ValueBitOr => ArithBitOr when we prove that both input are
definitely Int32. We also expand CompareEq etc.'s coverage onto the
proven types and some more additional handling in ArithBitOr in AI.

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):

Canonical link: https://commits.webkit.org/284721@main
